SmartDeviceLink/SDLSecondaryTransportManager.m

```diff
@@ -0,0 +1,715 @@
+
+    [self sdl_disconnectSecondaryTransport];
+}
+
+- (void)willTransitionFromStateRegisteredToStateReconnecting {
+    SDLLogD(@"Stopping services on secondary transport");
+    [self sdl_handleTransportUpdateWithPrimaryAvailable:YES secondaryAvailable:NO];
+
+    [self sdl_disconnectSecondaryTransport];
+}
+
+- (void)willTransitionFromStateRegisteredToStateStopped {
+    // sdl_handleTransportUpdateWithPrimaryAvailable is called in stop method
+    [self sdl_disconnectSecondaryTransport];
+}
+
+- (void)didEnterStateReconnecting {
+    __weak typeof(self) weakSelf = self;
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(RetryConnectionDelay * NSEC_PER_SEC)), _stateMachineQueue, ^{
+        if ([weakSelf.stateMachine isCurrentState:SDLSecondaryTransportStateReconnecting]) {
+            SDLLogD(@"Retry secondary transport after disconnection or registration failure");
+            [weakSelf.stateMachine transitionToState:SDLSecondaryTransportStateConfigured];
+        }
+    });
+}
+
+- (void)sdl_startManager {
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_onAppStateUpdated:) name:UIApplicationDidBecomeActiveNotification object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_onAppStateUpdated:) name:UIApplicationWillResignActiveNotification object:nil];
+
+    [self.primaryProtocolHandler start];
+}
+
+- (void)sdl_stopManager {
+    SDLLogD(@"SDLSecondaryTransportManager stop");
+
+    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidBecomeActiveNotification object:nil];
+    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationWillResignActiveNotification object:nil];
+
+    [self.primaryProtocolHandler stop];
+
+    self.streamingServiceTransportMap = [@{@(SDLServiceTypeAudio):@(SDLTransportClassInvalid),
+                                @(SDLServiceTypeVideo):@(SDLTransportClassInvalid)} mutableCopy];
+    self.secondaryTransportType = SDLSecondaryTransportTypeDisabled;
+    self.transportsForAudioService = @[];
+    self.transportsForVideoService = @[];
+
+    self.ipAddress = nil;
+    self.tcpPort = TCPPortUnspecified;
+}
+
+- (void)sdl_configureManager:(nullable NSArray<SDLSecondaryTransportTypeBox *> *)availableSecondaryTransports
+          availableTransportsForAudio:(nullable NSArray<SDLTransportClassBox *> *)availableTransportsForAudio
+          availableTransportsForVideo:(nullable NSArray<SDLTransportClassBox *> *)availableTransportsForVideo {
+    if (![self.stateMachine isCurrentState:SDLSecondaryTransportStateStarted]) {
+        SDLLogW(@"SecondaryTransportManager ignores duplicate Start Service ACK frame");
+        return;
+    }
+
+    SDLSecondaryTransportType secondaryTransportType = SDLSecondaryTransportTypeDisabled;
+    if (availableSecondaryTransports == nil || availableSecondaryTransports.count == 0) {
```

Comment:
`availableSecondaryTransports.count` will return `0` if it's `nil`. In this case, it's okay to only have `if(availableSecondaryTransports.count == 0)`

Reply:
Fixed in commit 068f8c142989fb8035fffed02f9c5d818b25d8cb.